File: Apps/Chat/android/chatkit/src/main/kotlin/com/dweekly/cyrinx/chat/ChatTransportClient.kt
package com.dweekly.cyrinx.chat

import kotlinx.coroutines.flow.Flow

/**
 * Platform-neutral chat transport client contract. ../../../CONTRACT.md
 * section 1.8. Implemented in this module by [SimulatedChatTransportClient]; the
 * live Cyrinx 3 SDK adapter is added in a later stage (C3-31) behind this same
 * interface.
 *
 * DECISION (not pinned by the brief, per ../../../CONTRACT.md section 1.8): the
 * [connect] parameter name (`peerIdHex`) matches the Swift argument label's
 * meaning (`toPeer idHex:`) since Kotlin has no external/internal label split;
 * error signaling uses plain thrown exceptions on `suspend fun`s (Kotlin has no
 * `async throws` distinction) rather than a `Result`-wrapping return type,
 * matching this repo's existing Kotlin conventions (no `Result<T>` usage found in
 * Apps/HIL/android).
 */
interface ChatTransportClient {
    suspend fun start()

    suspend fun stop()

    /** Bounded buffer ([ChatEventBus.EVENT_BUFFER_CAPACITY], drop-oldest); see
     * [ChatEvent]'s eventSeq gap-detection contract. */
    val events: Flow<ChatEvent>

    suspend fun connect(peerIdHex: String)

    suspend fun disconnect()

    /**
     * Returns on queue acceptance, NOT delivery -- see
     * [ChatMessageDisplayStatus]'s honest-delivery caveat. The returned string is
     * the new message's `messageIdHex`.
     */
    suspend fun send(body: String): String

    suspend fun cancelSend(messageIdHex: String)
}

File: Apps/Chat/CyrinxChatKit/Sources/CyrinxChatKit/ChatTransportClient.swift
import Foundation

/// Platform-neutral chat transport contract. Apps/Chat/CONTRACT.md §1.8.
///
/// Consumed by UI models (never touched directly by UI code -- see
/// Apps/Chat/README.md); implemented here by `SimulatedChatTransportClient`
/// and, in C3-31, by a live Cyrinx 3 SDK adapter behind the same contract.
public protocol ChatTransportClient {
    func start() async throws
    func stop() async

    /// Bounded, buffering newest 512 (drop-oldest); see the `eventSeq`
    /// gap-detection contract and bounded-buffer policy documented on
    /// `ChatEvent` and implemented by `ChatEventEmitter`
    /// (CONTRACT.md §1.7): a consumer detects drops by watching for
    /// `eventSeq` gaps (`gap = currentEventSeq - previousEventSeq - 1`);
    /// survivors keep their true, originally assigned `eventSeq` and
    /// relative order, so a gap is always detectable even though it is
    /// never actively resynced in this sample.
    var events: AsyncStream<ChatEvent> { get }

    func connect(toPeer idHex: String) async throws
    func disconnect() async

    /// Returns on queue acceptance, NOT delivery -- see
    /// `ChatMessageDisplayStatus`'s honest-delivery caveat (CONTRACT.md
    /// §1.5). The returned string is the new message's `messageIdHex`.
    func send(body: String) async throws -> String
    func cancelSend(messageIdHex: String) async
}
